subprojects/plugins/src/main/java/org/gradle/api/plugins/JavaTestFixturesPlugin.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright 2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.gradle.api.plugins;
+
+import org.gradle.api.Incubating;
+import org.gradle.api.Plugin;
+import org.gradle.api.Project;
+import org.gradle.api.artifacts.ProjectDependency;
+import org.gradle.api.artifacts.dsl.DependencyHandler;
+import org.gradle.api.tasks.SourceSet;
+import org.gradle.internal.component.external.model.ProjectTestFixtures;
+
+import static org.gradle.api.plugins.internal.JavaPluginsHelper.addApiToSourceSet;
+import static org.gradle.internal.component.external.model.TestFixturesSupport.TEST_FIXTURES_API;
+import static org.gradle.internal.component.external.model.TestFixturesSupport.TEST_FIXTURES_FEATURE_NAME;
+import static org.gradle.internal.component.external.model.TestFixturesSupport.TEST_FIXTURE_SOURCESET_NAME;
+
+/**
+ * Adds support for producing test fixtures. This plugin will automatically
+ * create a `testFixtures` source set, and wires the tests to use those
+ * test fixtures automatically.
+ *
+ * Other projects may consume the test fixtures of the current project by
+ * declaring a dependency using the {@link DependencyHandler#testFixtures(Object)}
+ * method.
+ *
+ * @since 5.6
+ */
+@Incubating
+public class JavaTestFixturesPlugin implements Plugin<Project> {
+
+    @Override
+    public void apply(Project project) {
+        project.getPluginManager().withPlugin("java", plugin -> {
```

Comment:
I wonder if we should not be even more opinionated here and have this plugin automatically apply the `java-library` plugin. After all, when you add separated test fixtures, you are building a library.
Wdyt?

Reply:
I don't think so. First, there are a few drawbacks in using `java-library` (performance on Windows, cost of migration) and it still makes sense for folks using `java` to produce test fixtures. But, more importantly, I can see that you would need to publish test fixtures for applications too. It's not uncommon, say, you have a CLI that uses the test fixtures of a server. The server is _not_ a library, but there are still test fixtures to use it.